### ISSUE TITLE
[SECURITY] SSH exposed with `root:root`

### ISSUE THREAD

**Issue by ValdikSS** [sipeed/NanoKVM#198]
NanoKVM [v1.3.0 image](https://github.com/sipeed/NanoKVM/releases/tag/v1.3.0) (20241120_NanoKVM_Rev1_3_0.img.xz) comes with SSH server which accepts connections from everywhere with `root:root` credentials.

This will result in a network breach in a matter of 15 minutes if anyone connect NanoKVM to the internet with a dedicated IPv4 address.

```
# ss -nlpt | grep 22
LISTEN 0      0            0.0.0.0:22        0.0.0.0:*    users:(("sshd",pid=470,fd=3))           
LISTEN 0      0                  *:22              *:*    users:(("sshd",pid=470,fd=4))
```

```
# iptables -vnL INPUT
Chain INPUT (policy ACCEPT 3183 packets, 607K bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     6    --  eth0   *       0.0.0.0/0            0.0.0.0/0            tcp dpt:80 state NEW,ESTABLISHED
    0     0 ACCEPT     6    --  eth0   *       0.0.0.0/0            0.0.0.0/0            tcp spt:22 state ESTABLISHED

# ip6tables -vnL INPUT
Chain INPUT (policy ACCEPT 362 packets, 33994 bytes)
 pkts bytes target     prot opt in     out     source               destination
```

Moreover, when [IP routing is disabled](https://github.com/sipeed/NanoKVM/issues/197), the device assigns global IPv6 address in IPv6-enabled networks, which is used for NTP requests, and may expose the device to rogue NTP servers: [Using IPv6 with Linux? You’ve likely been visited by Shodan and other scanners](https://arstechnica.com/information-technology/2016/02/using-ipv6-with-linux-youve-likely-been-visited-by-shodan-and-other-scanners/).

Such setup is very common for residential networks (contrary to dedicated IPv4), and may result in a network breach on a large scale.

### Mitigations

_Please follow [IoT Device Security Specification 1.0](https://csa-iot.org/newsroom/the-connectivity-standards-alliance-product-security-working-group-launches-the-iot-device-security-specification-1-0/) ([PDF](https://csa-iot.org/wp-content/uploads/2024/03/23-80986-013-PSWG-1.0-Specification-18-March-2024.pdf)) to learn more about IoT device security best practices, design and checklists_

1. Disable SSH daemon by default, enable it with a checkbox in web interface settings, which are available only after authentication
2. Accept connections only from [LAN ranges](https://en.wikipedia.org/wiki/Reserved_IP_addresses) for both IPv4 and IPv6 by default, accept from everywhere with additional checkbox in the settings
3. Do not use hardcoded `root` credential, but generate unique strong root password and show it on OLED display when SSH checkbox is enabled

### Workaround

1. Login via SSH as `root:root`
2. Set different root password with `passwd` command

**Comment by CRCinAU** [comment]
It's also worth mentioning that you should never, *never* put devices like a KVM with open, unfiltered inbound connectivity.

This includes cameras, printers, and just about every other device you would connect to a network.

**Comment by schlypel** [comment]
Ooh, this is bad, my router gives global ipv6 to all devices, if my ISP gives me a /64, which they do since a year or so...
Tried to access the ipv6 of nanoKVM, root:root got me in. No clue how many bots accessed before me...

```
# cat /etc/sysctl.conf
net.ipv6.conf.all.disable_ipv6=1
net.ipv6.conf.default.disable_ipv6=1
```

Creating the sysctl.conf with those 2 lines disabled all ipv6 functionality on my nanoKVM, lets hope I'm not part of botnet allready.

**Comment by Michal4K** [comment]
> No clue how many bots accessed before me...

On IPv6 it's not that big deal. The address space is so big that scanning it without hints (DNS...) is practically impossible. But this change nothing on that you should have properly setup ipv6 FW on your router.

**Comment by schlypel** [comment]
You are absolutely correct, no internal system should be visible through the firewall of my router.

But ISP provided routers frequently come with suboptimal defaults. Other users might wan't to:
- check their ipv6 connectivity
- change the root password
- set up ssh keys for access as root
- and disable logins with password.

**Comment by ValdikSS** [comment]
No network device should come with unmentioned remote control function which is accessible using widely well-known credentials. Blaming the user for not treating the device as a threat out of the box is not an rational answer, especially for IP-KVM, which is made to be externally accessible.

The device should follow [IoT Device Security Specification 1.0](https://csa-iot.org/newsroom/the-connectivity-standards-alliance-product-security-working-group-launches-the-iot-device-security-specification-1-0/) ([PDF](https://csa-iot.org/wp-content/uploads/2024/03/23-80986-013-PSWG-1.0-Specification-18-March-2024.pdf)).

Please refrain from commenting in this issue. This ticket is for manufacturer, not for discussion. The first post already has a workaround for the user.

**Comment by CRCinAU** [comment]
> Blaming the user for not treating the device as a threat out of the box is not an rational answer, especially for IP-KVM, which is made to be externally accessible.

Come on mate, back up a bit.... *anybody* placing an IP-KVM - no matter the vendor - on a publicly accessible, none filtered connection has already violated so many best practices, the username and password are so far down the list to not matter...

> The device should follow IoT Device Security Specification 1.0 (PDF).

A KVM is not an IoT device. Unless you're going to extend the definition of IoT to include just about anything that can connect to a network.

> Please refrain from commenting in this issue. This ticket is for manufacturer, not for discussion.

I don't think you get to dictate the input on a public forum.

**Comment by taitix** [comment]
I'm not a network savvy (especially ipv6) , but does this meam whoever observe a packet to the firmware update server can just look at the source ipv6 address and ssh into it?

**Comment by CRCinAU** [comment]
> I'm not a network savvy (especially ipv6) , but does this mean whoever observes a packet to the firmware update server can just look at the source ipv6 address and ssh into it?

No. What it means is that if you have the device open to the internet with no firewall on any router etc, that if you don't manually change the password, people would be able to log in with the default username and password.

This is the same on both IPv4 and IPv6 networks.

**Comment by ZeChArtiahSaher** [comment]
just turn it off by default, this shouldn't be a debate

**Comment by klosz007** [comment]
Changing of root's password should be enforced by GUI, like it is with admin's password.

Additionally SSH should be disabled by default and there should be GUI option to enable it.

Leaving SSH enabled with direct root access enabled via SSH and simple root's password is extremely bad practice, even for home environments.
At current state, from security perspective, this device is just a joke and usable only at home labs (and even this after some basic hardening).

**Comment by klosz007** [comment]
Weak root login over physical serial console (only) to allow reset of admin's password would be fine - it would require physical access to device to leverage such 'vulnerability'. But device has so little configurable options that resetting forgotten admin's password via just wiping off all settings via reset to factory defaults (or even by reflashing SD card) would also be an acceptable option here. Giving 'easy' root access over network just to allow reset of admin's password via network is just a big security hole. The device is great because it works and is cheap but from security point of view there's still a lot to be done to make it usable beyond expierimental/lab environments.

**Comment by woolmonkey** [comment]
> No network device should come with unmentioned remote control function which is accessible using widely well-known credentials. Blaming the user for not treating the device as a threat out of the box is not an rational answer, especially for IP-KVM, which is made to be externally accessible.
> 
> The device should follow [IoT Device Security Specification 1.0](https://csa-iot.org/newsroom/the-connectivity-standards-alliance-product-security-working-group-launches-the-iot-device-security-specification-1-0/) ([PDF](https://csa-iot.org/wp-content/uploads/2024/03/23-80986-013-PSWG-1.0-Specification-18-March-2024.pdf)).
> 
> Please refrain from commenting in this issue. This ticket is for manufacturer, not for discussion. The first post already has a workaround for the user.
> 
> At least as of firmware v1.3.0, the device does not assign global IPv6, because it's a router #197, which disables Router Advertisement by default.

And it wouldn't matter a global ipv6 does not mean it is on the net.  You still have to allow your firewall to allow traffic through.

**Comment by jerome-labidurie** [comment]
This seems to be a little bit mitigated with the 2.1.4 version of the application https://github.com/sipeed/NanoKVM/commit/d7ca7c453d821ad099bf79b463969419041279cb

sshd is still running, root can still login, but changing the user password via the GUI also change the root password.

**Comment by ValdikSS** [comment]
Let's call it fixed then, although I'd prefer the checkbox do disable ssh as well.

**Comment by ValdikSS** [comment]
Let's tag @Z2Z-GuGu then, he's responsible for firmware releases as far as I understand.

**Comment by ValdikSS** [comment]
Pinged Sipeed on Twitter once again.
https://x.com/ValdikSS/status/1887242397077090582

**Comment by ValdikSS** [comment]
@wj-xiao ^^^

**Comment by apalrd** [comment]
> https://www.youtube.com/watch?v=plJGZQ35Q6I

yup that's me

**Comment by wj-xiao** [comment]
In the latest version, users will be prompted to change the default password. 
The new password will apply to both the web login password and the root user password.

In the next version, SSH will be disabled by default. And users can enable or disable SSH through the web ui.

**Comment by wj-xiao** [comment]
SSH is disabled by default in image version [v1.4.0](https://github.com/sipeed/NanoKVM/releases/tag/v1.4.0).

In application version [2.2.0](https://github.com/sipeed/NanoKVM/releases/tag/2.2.0), SSH can be enabled or disabled via the web settings.
